src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala

```diff
@@ -0,0 +1,183 @@
+  * lost task.
+  *
+  * For each kill request, a child [[TaskKillProgressActor]] will be spawned, which
+  * is supposed to watch the progress and complete a given promise when all watched
+  * tasks are reportedly terminal.
+  *
+  * See [[TaskKillConfig]] for configuration options.
+  */
+private[impl] class TaskKillServiceActor(
+    taskTracker: TaskTracker,
+    driverHolder: MarathonSchedulerDriverHolder,
+    stateOpProcessor: TaskStateOpProcessor,
+    config: TaskKillConfig,
+    clock: Clock) extends Actor with ActorLogging {
+  import TaskKillServiceActor._
+  import context.dispatcher
+
+  val tasksToKill: mutable.HashMap[Task.Id, Option[Task]] = mutable.HashMap.empty
+  val inFlight: mutable.HashMap[Task.Id, TaskToKill] = mutable.HashMap.empty
+  var retryTimer: Option[Cancellable] = None
+
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[MesosStatusUpdateEvent])
+  }
+
+  override def receive: Receive = {
+    case KillTaskById(taskId, promise) =>
+      killTaskById(taskId, promise)
+
+    case KillUnknownTaskById(taskId, promise) =>
+      killUnknownTaskById(taskId, promise)
+
+    case KillTasks(tasks, promise) =>
+      killTasks(tasks, promise)
+
+    case Terminal(event) if inFlight.contains(event.taskId) || tasksToKill.contains(event.taskId) =>
+      handleTerminal(event.taskId)
+
+    case Retry =>
+      retry()
+
+    case unhandled: InternalRequest =>
+      log.warning("Received unhandled {}", unhandled)
+  }
+
+  def killTaskById(taskId: Task.Id, promise: Promise[Done]): Unit = {
+    log.debug("Received KillTaskById({})", taskId)
+    import context.dispatcher
+    taskTracker.task(taskId).map {
+      case Some(task) => self ! KillTasks(Seq(task), promise)
+      case _ => self ! KillUnknownTaskById(taskId, promise)
+    }
+  }
+
+  def killUnknownTaskById(taskId: Task.Id, promise: Promise[Done]): Unit = {
+    log.debug("Received KillUnknownTaskById({})", taskId)
+    setupProgressActor(Seq(taskId), promise)
+    tasksToKill.update(taskId, None)
+    processKills()
+  }
+
+  def killTasks(tasks: Iterable[Task], promise: Promise[Done]): Unit = {
+    log.debug("Adding {} tasks to queue; setting up child actor to track progress", tasks.size)
+    setupProgressActor(tasks.map(_.taskId), promise)
+    tasks.foreach { task =>
+      tasksToKill.update(task.taskId, Some(task))
+    }
+    processKills()
+  }
+
+  def setupProgressActor(taskIds: Iterable[Task.Id], promise: Promise[Done]): Unit = {
+    context.actorOf(TaskKillProgressActor.props(taskIds, promise))
+  }
+
+  def processKills(): Unit = {
+    val killCount = config.killChunkSize - inFlight.size
+    val toKillNow = tasksToKill.take(killCount)
+
+    log.info("processing {} kills", toKillNow.size)
+    toKillNow.foreach {
+      case (taskId, maybeTask) => processKill(taskId, maybeTask)
+    }
+
+    retryTimer = Some(context.system.scheduler.scheduleOnce(config.killRetryTimeout, self, Retry))
```

Comment:
You call `processKills` in different methods. 
This method schedules a `Retry` message. (Also `retry` schedules a `Retry` message).
To not schedule more messages than needed, either look for already scheduled messages or cancel the already schedules message.